File: src/ts/0977_squares_of_a_sorted_array.ts
function sortedSquares(nums: number[]): number[] {
    const res: number[] = [];
    let left = 0,
        right = nums.length - 1;

    while (left <= right) {
        if (Math.abs(nums[left]) > nums[right]) {
            res.unshift(nums[left] ** 2);
            left++;
        } else {
            res.unshift(nums[right] ** 2);
            right--;
        }
    }

    return res;
}

const nums = [-4, -1, 0, 3, 10];
console.log(sortedSquares(nums)); // [0, 1, 9, 16, 100]

File: leetpattern/python/0901_1200/0977_squares_of_a_sorted_array.py
from typing import List


# Left Right Pointers
def sortedSquares(nums: List[int]) -> List[int]:
    """Returns the squares of the sorted array."""
    n = len(nums)
    result = [0 for _ in range(n)]

    left, right, tail = 0, n - 1, n - 1

    while left <= right:
        if abs(nums[left]) >= abs(nums[right]):
            result[tail] = nums[left] ** 2
            left += 1
        else:
            result[tail] = nums[right] ** 2
            right -= 1
        tail -= 1

    return result


# |---------------------|------|-------|
# | Approach            | Time | Space |
# |---------------------|------|-------|
# | Left Right Pointers | O(n) | O(n)  |
# |---------------------|------|-------|


nums = [-4, -1, 0, 3, 10]
print(sortedSquares(nums))  # [0, 1, 9, 16, 100]
